crates/asm/common/src/aux.rs

```diff
@@ -1,56 +1,136 @@
+/// Responsible for recording auxiliary requests emitted during preprocessing.
+pub trait AuxInputCollector: Any {
+    /// Records that the transaction at `tx_index` requires auxiliary data described by `payload`.
+    fn request_aux_input(&mut self, tx_index: L1TxIndex, payload: Box<dyn AuxRequestPayload>);
+
+    /// Exposes the collector as a `&mut dyn Any` for downcasting.
+    fn as_mut_any(&mut self) -> &mut dyn Any;
+}
+
+/// Extension trait adding ergonomic helpers for [`AuxInputCollector`] implementers.
+pub trait AuxInputCollectorExt: AuxInputCollector {
+    /// Convenience helper that accepts concrete payload types without manual boxing.
+    fn request_aux<T>(&mut self, tx_index: L1TxIndex, payload: T)
+    where
+        T: AuxRequestPayload,
+    {
+        self.request_aux_input(tx_index, Box::new(payload));
     }
 }
+
+impl<T: AuxInputCollector + ?Sized> AuxInputCollectorExt for T {}
+
+/// Errors that can occur while resolving auxiliary data.
+#[derive(Debug, Error)]
+pub enum AuxResolveError {
+    /// The available aux data does not match the expected variant.
+    #[error(
+        "unexpected auxiliary response for subprotocol {subprotocol}, tx index {tx_index} (expected {expected:?}, provided {actual:?})"
+    )]
+    UnexpectedResponseVariant {
+        /// Subprotocol identifier.
+        subprotocol: SubprotocolId,
+        /// L1 transaction index within the block.
+        tx_index: L1TxIndex,
+        /// Expected variant.
+        expected: AuxResponseKind,
+        /// Response that was provided.
+        actual: AuxResponseKind,
+    },
+    /// Verification of the supplied MMR proof failed.
+    #[error(
+        "log MMR verification failed for subprotocol {subprotocol}, tx index {tx_index}, block {block_hash}"
+    )]
+    InvalidLogProof {
+        /// Subprotocol identifier.
+        subprotocol: SubprotocolId,
+        /// L1 transaction index within the block.
+        tx_index: L1TxIndex,
+        /// Hash of the L1 block whose logs were being proven.
+        block_hash: BlockHash,
+    },
+}
+
+/// Result alias for aux resolution operations.
+pub type AuxResolveResult<T> = Result<T, AuxResolveError>;
+
+/// Enumerates the different auxiliary response variants.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum AuxResponseKind {
+    HistoricalLogs,
+    HistoricalLogsRange,
+    DepositRequestTx,
+}
+
+/// Provides access to pre-fetched auxiliary responses during transaction execution.
+pub trait AuxResolver {
+    /// Returns all verified historical ASM logs associated with a given L1 transaction.
+    fn historical_logs(&self, tx_index: L1TxIndex) -> AuxResolveResult<Vec<AsmLogEntry>>;
+
+    /// Returns the deposit request transaction referenced by a bridge deposit, if present.
+    // TODO: consider changing return type based on bridge subprotocol needs
+    fn deposit_request_tx(&self, tx_index: L1TxIndex) -> AuxResolveResult<Option<Vec<u8>>>;
+}
```

Comment:
Similarly, if we're saying that "the aux has to be one of these types" then this adds indirection and we can just have an enum for the entries that would be.